src/execution_plan/ops/op_unwind.c

```diff
@@ -32,37 +37,87 @@ OpBase *NewUnwindOp(uint record_idx, AR_ExpNode **exps) {
 }
 
 OpResult UnwindInit(OpBase *opBase) {
+	OpUnwind *unwind = (OpUnwind *)opBase;
+
+	// check for modifiers in the AR_EXP for static or dynamic list
+	rax *modifiers = raxNew();
+	AR_EXP_CollectEntityIDs(unwind->expressions, modifiers);
+	// if there aren't any modifiers - list is static
+	if(!raxSize(modifiers)) {
+		// set the list
+		unwind->list = AR_EXP_Evaluate(unwind->expressions, NULL);
+		assert(unwind->list.type == T_ARRAY);
+		unwind->isStatic = true;
+	}
+	// check if there are children - if not, add static record and reset the list index
+	if(!unwind->op.childCount) {
+		unwind->currentRecord = Record_New(opBase->record_map->record_len);
+		// the list index is set to 0 only when there are no children to the op
+		// if the list is static, but the op has children, we first need to get a record from them
+		// MATCH(n) UNWIND [0,1,2] as y return n, y
+		unwind->listIdx = 0;
+	}
 	return OP_OK;
 }
 
+// try to generate new value to return
+// NULL will be returned if dynamic list is not evaluted (listIdx = UNDEFIND)
+// or in case of the current list fully returned its memebers
+Record _handoff(OpUnwind *op) {
+	// if there is a new value ready, return it
+	if(op->listIdx < SIArray_Length(op->list)) {
+		Record r = Record_Clone(op->currentRecord);
+		Record_AddScalar(r, op->unwindRecIdx, SIArray_Get(op->list, op->listIdx));
+		op->listIdx++;
+		return r;
+	}
+	return NULL;
+}
+
 Record UnwindConsume(OpBase *opBase) {
 	OpUnwind *op = (OpUnwind *)opBase;
 
-	// Evaluated and returned all expressions.
-	if(op->expIdx == array_len(op->expressions)) return NULL;
+	// check for new value
+	Record res = _handoff(op);
+	if(res) return res;
 
-	AR_ExpNode *exp = op->expressions[op->expIdx];
-	Record r = Record_New(opBase->record_map->record_len);
-	SIValue v = AR_EXP_Evaluate(exp, r);
-	Record_AddScalar(r, op->unwindRecIdx, v);
-	op->expIdx++;
+	// no new value - check if there are new lists to unwind
+	// check if dynamic or static unwind
+	if(op->op.childCount) {
+		Record r;
+		OpBase *child = op->op.children[0];
+		// if there are new lists to unwind
+		if((r = OpBase_Consume(child))) {
+			if(op->listIdx != INDEX_NOT_SET) Record_Free(op->currentRecord);
+			op->currentRecord = r;
 
-	return r;
+			// TODO: if expression is static or not
+			if(!op->isStatic) {
+				SIValue_Free(&op->list);
+				// set the list
+				op->list = AR_EXP_Evaluate(op->expressions, r);
+				assert(op->list.type == T_ARRAY);
+			}
+			// resed index
+			op->listIdx = 0;
+		}
+	}
+	// in case of no more lists from the op's children, or p has no children and static list returned fully, _handoff(op) will return NULL
+	return _handoff(op);
 }
 
 OpResult UnwindReset(OpBase *ctx) {
 	OpUnwind *unwind = (OpUnwind *)ctx;
-	unwind->expIdx = 0;
+	unwind->listIdx = INDEX_NOT_SET;
```

Comment:
What about freeing the list if it is not static?
Another idea would be, in case of a static list, modify/replace the arithmetic expression with one which simply holds a const value (the array) this way the logic will be exactly the same and we can remove the `isStatic` varaible
lastly who's responsible for freeing the expression?

Reply:
the list will be freed once the consume op will be called again